docs/content/best-practices/modeling.mdx

```diff
@@ -0,0 +1,176 @@
+
+```dsl.openfga
+
+  type user
+
+  type entity
+    relations
+      define admin: [user]
+      define billing_manager: [user]
+
+      define can_manage_billing: admin or billing_manager
+      define can_manage_users: admin
+```
+
+Defining the model in a way that closely resembles your application domain has several advantages:
+
+  - **Enhanced clarity and maintainability**: Authorization logic is easier to understand, debug, and maintain. By just reading the model, developers or security auditors can readily grasp the meaning of each type and relationship.
+
+  - **Better performance**: Models with more specific types and flatter hierarchies often lead to better performance. This is because <ProductName format={ProductNameFormat.ShortForm}/> can more efficiently process queries involving well-defined types and relationships compared to navigating complex, high-cardinality recursive relationships within a single generic type.
+
+  - **Leveraging the model's flexibility**: <ProductName format={ProductNameFormat.ShortForm}/>'s modeling language is designed to be adaptable. You can define numerous distinct types and relationships without significant overhead. Model changes rarely require data migrations, allowing you to evolve your model as your application grows. <ProductName format={ProductNameFormat.ShortForm}/> is already a generic authorization engine, you don't need to build another one on top of it.
+
+  - **Improved Collaboration**: The resource types that are owned by each application team can be maintained in independent [modules](./../modeling/modular-models.mdx). You can then control which application can write to specific resource types. For example, when using the API credential issued to a Document Management application, you can only write/delete tuples for documents and folders, but not for other types. This is not possible if you use a generic `resource` type, for example. 
+
+## Custom Roles
+
+However, in some applications, you will want end-users to define their own roles. In that case, you do need a role type:
+
+```dsl.openfga
+ model
+    schema 1.1 
+ type user
+
+  type role
+   relations
+     define assignee: [user]
+
+  type entity
+    relations
+      define admin: [user]
+      define billing_manager: [user]
+
+      define can_manage_billing: [role#assignee] or admin or billing_manager
+      define can_manage_users: [role#assignee] or admin
+```
+
+In this example we are combining static roles with dynamic roles. This is the recommended way of doing it. Do not always use generic roles. Define static roles for the ones you already know, and implement generic roles if your application needs them. Adding additional static roles is very easy, you just need to add a relation in the model, and it does not happen often. 
+
+You can see a full example of implementing custom roles [here](https://github.com/openfga/sample-stores/tree/main/stores/custom-roles).
```

Comment:
Note to future self: We probably want to update the custom roles sample and the [docs](https://openfga.dev/docs/modeling/custom-roles) page with some of the stuff mentioned here.